sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSqlUtils.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.zetasql;
+
+import com.google.protobuf.ByteString;
+import com.google.zetasql.ArrayType;
+import com.google.zetasql.StructType;
+import com.google.zetasql.StructType.StructField;
+import com.google.zetasql.Type;
+import com.google.zetasql.TypeFactory;
+import com.google.zetasql.Value;
+import com.google.zetasql.ZetaSQLType.TypeKind;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.Schema.Field;
+import org.apache.beam.sdk.schemas.Schema.FieldType;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.math.LongMath;
+import org.joda.time.Instant;
+
+/** Utility methods for ZetaSQL related operations. */
+@Internal
+public final class ZetaSqlUtils {
+
+  private static final long MICROS_PER_MILLI = 1000L;
+
+  private ZetaSqlUtils() {}
+
+  // Unsupported ZetaSQL types: INT32, UINT32, UINT64, FLOAT, ENUM, PROTO, GEOGRAPHY
+  // TODO[BEAM-8630]: support ZetaSQL types: DATE, TIME, DATETIME
+  public static Type beamFieldTypeToZetaSqlType(FieldType fieldType) {
+    switch (fieldType.getTypeName()) {
+      case INT64:
+        return TypeFactory.createSimpleType(TypeKind.TYPE_INT64);
+      case DECIMAL:
+        return TypeFactory.createSimpleType(TypeKind.TYPE_NUMERIC);
+      case DOUBLE:
+        return TypeFactory.createSimpleType(TypeKind.TYPE_DOUBLE);
+      case STRING:
+        return TypeFactory.createSimpleType(TypeKind.TYPE_STRING);
+      case DATETIME:
+        // TODO[BEAM-8630]: Mapping Timestamp to DATETIME results in some timezone/precision issues.
```

Comment:
We determined the timezone issue is non-existent. I wonder if we could make a logical type that gave us an extra field to stuff nanoseconds without breaking window functions?

Reply:
That should be possible, and I will work on that next. But I want to put all date/time type related change in a separate PR. SGTY?

Reply:
Sounds good.